### PR TITLE
fix(openapi-parser): rewrite parser

### PR DESCRIPTION
**Pull request by fuma-nama**
**Problem**
Currently, it uses `JSON.stringify` and requires multiple iteration.

**Solution**
With this PR, only one is needed:

The `resolve` function iterates over objects and resolve references, the resolved value will also be dereferenced to ensure nested references work. A `resolved` Set will act as a storage to detect if the object has been dereferenced, which replaces the need of `JSON.stringify` as self-referencing only happens with `$ref` in JSON schema.

### CODE REVIEW

File: packages/openapi-parser/src/utils/resolveReferences.ts [hunk 1]
```diff
@@ -105,73 +95,79 @@ export function resolveReferences(
...
+      for (const key of Object.keys(target)) {
+        if (schema[key] === undefined) {
+          schema[key] = target[key]
         }
       }
+    }
 
-      if (typeof value === 'object' && !isCircular(value)) {
-        result = resolve(value, resolveFilesystem, resolveFile, onResolve)
+    // Iterate over the whole objecct
```
Comment (hanspagel) [inline packages/openapi-parser/src/utils/resolveReferences.ts:144]:
```suggestion
    // Iterate over the whole object
```

**Review by hanspagel — CHANGES_REQUESTED** [review]
OMG, this is sick!!

I think the Git history is off, though. It seems to delete the changes of your last PR.

Do you want me to fix it?

File: packages/openapi-parser/src/utils/resolveReferences.ts (hunk 1 above)
Comment (fuma-nama) [reply packages/openapi-parser/src/utils/resolveReferences.ts]: Oops I shouldn't code at night ;)

**Comment by fuma-nama** [comment]
Oh I only moved that line to 132 and referenced it from `options` instead of adding parameter, it should continue to work

**Comment by hanspagel** [comment]
Really great PR, some benchmarking results:

  | Test Case        | Before                                        | After                                         |
  | ---------------- | --------------------------------------------- | --------------------------------------------- |
  | No Reference     | 3.73x faster than @apidevtools/swagger-parser | 5.25x faster than @apidevtools/swagger-parser |
  | Petstore         | 1.90x slower than @scalar/openapi-parser      | 2.23x faster than @apidevtools/swagger-parser |
  | Single Reference | 1.19x faster than @apidevtools/swagger-parser | 2.34x faster than @apidevtools/swagger-parser |

**Review by hanspagel — APPROVED** [review]
PR of the month 🏆

**Pull request merged**